File: third-semester/Data-structure-and-algorithms-II/backtracking/n-queens/n_queens.py
class n_queens():
    
    def __init__(self, N):
        self.N = N
        self.chess_board =  [[0 for i in range(N)] for j in range(N)]
        self.solution = 0

    def setQueen(self, row):
        if row == self.N:
            self.solution+=1
            print("Solution: " + str(self.solution))
            self.print_chess_board()
            return
        
        for col in range(self.N):
            if self.validate(row, col):
                self.chess_board[row][col] = 1
                self.setQueen(row+1)
                self.chess_board[row][col] = 0

    def validate(self, row, col):
        for r in range(row):
            if self.chess_board[r][col] == 1: return False

        r = row - 1
        c = col - 1
        while(r >= 0 and c >= 0):
            if self.chess_board[r][c] == 1: return False
            r -= 1
            c -= 1

        r = row - 1
        c = col + 1
        while(r >= 0  and c < self.N):
            if self.chess_board[r][c] == 1: return False
            r -= 1
            c += 1
        
        return True

    def print_chess_board(self):
        for row in range(self.N):
            for col in range(self.N):
                print(self.chess_board[row][col], end=" ")
            print()
        print()


def main():
    chess_board1 = n_queens(8)
    chess_board1.print_chess_board()
    chess_board1.setQueen(0)


if __name__ == '__main__':
    main()

File: 3-semester/Data-structure-and-algorithms-II/backtracking/n-queens/cpp/nQueens.cpp
#include <iostream>
#include <string>
#include "nQueens.h"
#include <vector>

using namespace std;

nQueens::nQueens(int N){
    this->N = N;
    this->board = vector<vector<int> >(N, vector<int>(N));
    //fill the board with 0
    for(int row = 0; row < N; row++){
        for(int col = 0; col < N; col++){
            this->board[row][col] = 0;
        }
    }
    this->solution = 0;
}

void nQueens::setQueens(int row){
    if(row == this->N) {
        this->solution++;
        cout << "Solution " + this->solution << endl;
        printBoard();
        return;
    }

    for(int col = 0; col < N; col++){
        if(validate(row, col)){
            this->board[row][col] = 1;
            setQueens(row+1);
            this->board[row][col] = 0;
        }
    }
}

bool nQueens::validate(int row, int col){
    // Validate row
    for(int r = 0; r < row; r++){
        if(this->board[r][col] == 1){
            return false;
        }
    }

    // Validate diagonal 1
    int r = row - 1; int c = col - 1;
    while(r >= 0 && c >= 0) {
        if(this->board[r][c] == 1){
            return false;
        }
        r--;
        c--;
    }

    // Validate diagonal 2
    r = row - 1; c = col + 1;
    while(r >= 0 && c >= 0) {
        if(this->board[r][c] == 1){
            return false;
        }
        r--;
        c++;
    }
    return true;
}

void nQueens::printBoard(){
    for(int row = 0; row < N; row++){
        for(int col = 0; col < N; col++){
            cout << this->board[row][col] + " ";
        }
        cout << endl;
    }
    cout << endl;
}

int main(){
    nQueens* chessBoard1 = new nQueens(8);
    chessBoard1->printBoard();
    chessBoard1->setQueens(0);
}
